### PR TITLE
Feature #6 - As user I want to delete a Bike

### LINKED ISSUE #6: As user I want to delete a Bike
"Hell NOOO!!! This is ****** DELETE IT!!!"

- Create a DELETE-Endpoint "/api/bikes/{id}"
- In the frontend I want to delete a Bike with just one click
  - Make a Button on the Card Component
  - Call the DELETE-Endpoint to delete the Bike
- write unit- and integrationtest

### PR DESCRIPTION
**Pull request by JohannesThorbergsson**

### CODE REVIEW

File: backend/src/main/java/com/bikes/backend/controller/BikeController.java [hunk 1]
```diff
@@ -3,23 +3,33 @@
...
+    }
+    @PostMapping("/")
+    public Bike addBike(@RequestBody Bike bikeToAdd) {
+        return bikeService.addBike(bikeToAdd);
+    }
 
+    @DeleteMapping("/{id}")
+    public Bike deleteBike(@PathVariable String id){
+        return bikeService.deleteBike(id);
+    }
```
Comment (Anmavel) [inline backend/src/main/java/com/bikes/backend/controller/BikeController.java:34]: Please only add one feature at the time for PR :-)

File: backend/src/main/java/com/bikes/backend/repository/BikeRepository.java [hunk 2]
```diff
@@ -2,22 +2,32 @@
...
+	}
 
+
+    public Bike addBike(Bike bikeToAdd) {
+        bikeMap.put(bikeToAdd.id(), bikeToAdd);
+        return bikeToAdd;
+    }
+	public Optional<Bike> deleteBike(String id) {
+		return Optional.ofNullable(bikeMap.remove(id));
+	}
```
Comment (Anmavel) [inline backend/src/main/java/com/bikes/backend/repository/BikeRepository.java:32]: Optional is not necessary, remove returns null and that can be checked in the service

File: backend/src/main/java/com/bikes/backend/service/BikeService.java [hunk 3]
```diff
@@ -1,23 +1,36 @@
...
 import com.bikes.backend.repository.BikeRepository;
-import lombok.RequiredArgsConstructor;
 import lombok.Getter;
+import lombok.RequiredArgsConstructor;
 import org.springframework.stereotype.Service;
+
 import java.util.List;
 
 @RequiredArgsConstructor
 @Getter
```
Comment (Anmavel) [inline backend/src/main/java/com/bikes/backend/service/BikeService.java]: Getter-Methods are not necessary for the Service

File: backend/src/main/java/com/bikes/backend/service/BikeService.java [hunk 4]
```diff
@@ -1,23 +1,36 @@
...
+		return bikeRepository.getBikeById(id).orElseThrow(NoSuchBikeException::new);
+	}
 
+	public Bike addBike(Bike incomingBike) {
+		Bike bikeToAdd = new Bike(idService.generateId(), incomingBike.title());
+		return bikeRepository.addBike(bikeToAdd);
+	}
 
+	public Bike deleteBike(String id){
+		return bikeRepository.deleteBike(id).orElseThrow(NoSuchBikeException::new);
```
Comment (Anmavel) [inline backend/src/main/java/com/bikes/backend/service/BikeService.java:32]: you could check for null and throw an Exception avoiding the use of an extra Optional

File: backend/src/main/java/com/bikes/backend/service/NoSuchBikeException.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+import java.util.NoSuchElementException;
+
+@ResponseStatus(code = HttpStatus.NOT_FOUND, reason = "Bike not found.")
+public class NoSuchBikeException extends NoSuchElementException {
+
+}
```
Comment (Anmavel) [inline backend/src/main/java/com/bikes/backend/service/NoSuchBikeException.java:11]: Nice Idea :-), we will copy this!

File: backend/src/test/java/com/bikes/backend/controller/BikeControllerTest.java [hunk 6]
```diff
@@ -25,14 +24,106 @@ class BikeControllerTest {
...
+            bikeRepository.addBike(testBike);
+
+            //when
+            mockMvc.perform(MockMvcRequestBuilders.get("/api/bikes/" + testBike.id()))
+                    .andExpect(status().isOk())
+                    .andExpect(content().json(""" 
+                                            {
+                                                "title": "testBike"
+                                            }
+                                    """));
```
Comment (Anmavel) [inline backend/src/test/java/com/bikes/backend/controller/BikeControllerTest.java]: Bike Id is missing in Json

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java [hunk 7]
```diff
@@ -3,43 +3,120 @@
...
-    BikeService bikeService = new BikeService(bikeRepository);
+	BikeRepository bikeRepository = mock(BikeRepository.class);
+	IdService idService = mock(IdService.class);
+    BikeService bikeService = new BikeService(bikeRepository, idService);
 
-    Bike testBike = new Bike("testBike", "testId");
-    List<Bike> expectedBikes = List.of(testBike);
-    List<Bike> expectedBikesEmpty = new ArrayList<>();
+	Bike testBike = new Bike("testId", "testBike");
+	List<Bike> expectedBikes = List.of(testBike);
```
Comment (Anmavel) [inline backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java:23]: Watch out: List.of() creates an immutable List . Keep it in mind while testing...

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java [hunk 8]
```diff
@@ -3,43 +3,120 @@
...
+		@Test
+		@DisplayName("...returns all bikes if the repo is not empty")
+		void getAllBikes_returnsAllBikesIfTheRepoIsNotEmpty() {
+			//GIVEN
+			when(bikeRepository.getAllBikes()).thenReturn(expectedBikes);
+			//WHEN
+			List<Bike> actual = bikeService.getAllBikes();
+			//THEN
+			verify(bikeRepository).getAllBikes();
+			Assertions.assertEquals(expectedBikes, actual);
```
Comment (Anmavel) [inline backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java:53]: expectedBikes will be better placed before verify, so it is easier to read

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java [hunk 9]
```diff
@@ -3,43 +3,120 @@
...
+			Assertions.assertEquals(testBike, actual);
+		}
+
+		@Test
+		@DisplayName("...throws an exception if no bike with the given id exists")
+		void getBikeById_throwExceptionIfNoBikeWithTheGivenIdExists() {
+			//GIVEN
+			String invalidId = "invalidId";
+			//WHEN
+			when(bikeRepository.getBikeById(invalidId)).thenThrow(new NoSuchBikeException());
```
Comment (Anmavel) [inline backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java]: Attention! Repository should not throw an Exception but instead return an empty Optional

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java [hunk 10]
```diff
@@ -3,43 +3,120 @@
...
+			Bike actual = bikeService.deleteBike(testBike.id());
+			//THEN
+			verify(bikeRepository).deleteBike(testBike.id());
+			Assertions.assertEquals(testBike, actual);
+		}
+		@Test
+		@DisplayName("deleting invalid BikeId")
+		void deleteBikeIdInvalid() {
+			//GIVEN
+			when(bikeRepository.deleteBike(testBike.id())).thenThrow(new NoSuchBikeException());
```
Comment (Anmavel) [inline backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java]: same Problem: don´t throw an Exception in the repository

File: frontend/src/components/AddBike/AddBike.tsx [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import React, { ChangeEvent, FormEvent, useState } from 'react'
+import './AddBike.css'
+
+type AddBikeProps = {
+   addBike: (newBikeTitle: string) => void
```
Comment (Anmavel) [inline frontend/src/components/AddBike/AddBike.tsx:5]: addBike function could return a Promise, see next comment

File: frontend/src/components/AddBike/AddBike.tsx [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+}
+
+export default function AddBike(props: AddBikeProps) {
+   const [inputTitle, setInputTitle] = useState<string>('')
+
+   function handleSubmit(event: FormEvent<HTMLFormElement>) {
+      event.preventDefault()
+      props.addBike(inputTitle.trim())
+      setInputTitle('')
```
Comment (Anmavel) [inline frontend/src/components/AddBike/AddBike.tsx:14]: you could reset the inputTitle in a .then block

**Review by Anmavel — APPROVED** [review]
Please only one feature at the time

File: backend/src/main/java/com/bikes/backend/service/NoSuchBikeException.java (hunk 5 above)
Comment (gadmel) [reply backend/src/main/java/com/bikes/backend/service/NoSuchBikeException.java:11]: Anytime)

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java (hunk 7 above)
Comment (gadmel) [reply backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java:23]: Ok

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java (hunk 8 above)
Comment (gadmel) [reply backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java:53]: No

File: backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java (hunk 9 above)
Comment (gadmel) [reply backend/src/test/java/com/bikes/backend/service/BikeServiceTest.java]: You're right! Thanks, deleted the line

File: frontend/src/components/AddBike/AddBike.tsx (hunk 11 above)
Comment (gadmel) [reply frontend/src/components/AddBike/AddBike.tsx:5]: Yes, but we have discussed that and decided to resolve the promise in the respective service (see. bikesApiService)

File: frontend/src/components/AddBike/AddBike.tsx (hunk 12 above)
Comment (gadmel) [reply frontend/src/components/AddBike/AddBike.tsx:14]: addBike is not async that's why that would not work

**Pull request merged**